File: pkg/tenet/engine.go
package tenet

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"
)

// Run executes the schema logic for a given effective date.
// It evaluates the logic tree, computes derived state, and validates the document.
// Returns the transformed JSON with computed state, errors, and status.
//
// This is the "Transformer" - it takes raw input and returns a fully evaluated document.
// Panic-safe: recovers from any unexpected panic and returns it as an error.
func Run(jsonText string, date time.Time) (result string, err error) {
	defer func() {
		if r := recover(); r != nil {
			result = ""
			err = fmt.Errorf("internal error: %v", r)
		}
	}()

	// 1. Unmarshal
	var schema Schema
	if err := json.Unmarshal([]byte(jsonText), &schema); err != nil {
		return "", fmt.Errorf("unmarshal: %w", err)
	}

	if schema.Definitions == nil {
		schema.Definitions = make(map[string]*Definition)
	}

	// Initialize default visibility for definitions
	for _, def := range schema.Definitions {
		if def != nil && def.Visible == nil {
			t := true
			def.Visible = &t
		}
	}

	engine := NewEngine(&schema)

	// 2. Validate and select temporal branch, prune inactive rules
	if len(schema.TemporalMap) > 0 {
		engine.validateTemporalMap()
		branch := engine.selectBranch(date)
		if branch != nil {
			engine.prune(branch)
		}
	}

	// 3. Compute derived state (so logic tree can use derived values)
	engine.computeDerived()

	// 4. Evaluate logic tree
	engine.evaluateLogicTree()

	// 5. Re-compute derived state (in case logic modified inputs)
	engine.computeDerived()

	// 6. Validate
	engine.validateDefinitions()
	engine.checkAttestations()

	// 7. Determine status and attach errors
	schema.Errors = engine.errors
	schema.Status = engine.determineStatus()

	// 8. Marshal result
	return engine.marshal()
}

// Verify checks that a completed document (newJson) was correctly derived from a base schema.
// It simulates the user's journey by iteratively copying visible field values and re-running.
//
// Optional maxIterations parameter (default: 100) limits the replay iterations.
//
// This is the "Auditor" - it proves the transformation was legal by replaying the journey.
// Returns a structured VerifyResult with all issues found (not just the first).
// Panic-safe: recovers from any unexpected panic and returns it as an internal_error issue.
func Verify(newJson, baseSchemaJson string, maxIter ...int) (vr VerifyResult) {
	defer func() {
		if r := recover(); r != nil {
			vr = VerifyResult{
				Valid: false,
				Issues: []VerifyIssue{{
					Code:    VerifyInternalError,
					Message: fmt.Sprintf("internal panic: %v", r),
				}},
				Error: fmt.Sprintf("internal panic: %v", r),
			}
		}
	}()

	maxIterations := 100
	if len(maxIter) > 0 && maxIter[0] > 0 {
		maxIterations = maxIter[0]
	}

	// Parse both documents
	var newSchema Schema
	if err := json.Unmarshal([]byte(newJson), &newSchema); err != nil {
		return VerifyResult{
			Valid: false,
			Issues: []VerifyIssue{{
				Code:    VerifyInternalError,
				Message: fmt.Sprintf("failed to parse submitted document: %v", err),
			}},
			Error: fmt.Sprintf("unmarshal newJson: %v", err),
		}
	}

	// Extract effective date from newJson
	effectiveDate := time.Now()
	if newSchema.ValidFrom != "" {
		if parsed, ok := parseDate(newSchema.ValidFrom); ok {
			effectiveDate = parsed
		}
	}

	// Start with base schema
	currentJson := baseSchemaJson
	previousVisibleSet := ""

	for iteration := 0; iteration < maxIterations; iteration++ {
		// Parse current state
		var currentSchema Schema
		if err := json.Unmarshal([]byte(currentJson), &currentSchema); err != nil {
			return VerifyResult{
				Valid: false,
				Issues: []VerifyIssue{{
					Code:    VerifyInternalError,
					Message: fmt.Sprintf("failed to parse schema at iteration %d", iteration),
				}},
				Error: fmt.Sprintf("unmarshal current (iteration %d): %v", iteration, err),
			}
		}

		// Count visible editable fields before copying
		visibleEditable := getVisibleEditableFields(&currentSchema)

		// Copy values from newJson for visible, editable fields
		for fieldId := range visibleEditable {
			if newDef, ok := newSchema.Definitions[fieldId]; ok && newDef != nil {
				if currentDef, ok := currentSchema.Definitions[fieldId]; ok && currentDef != nil {
					currentDef.Value = newDef.Value
				}
			}
		}

		// Copy attestation states for visible attestations
		for attId, currentAtt := range currentSchema.Attestations {
			if currentAtt == nil {
				continue
			}
			if newAtt, ok := newSchema.Attestations[attId]; ok && newAtt != nil {
				currentAtt.Signed = newAtt.Signed
				currentAtt.Evidence = newAtt.Evidence
			}
		}

		// Run the schema
		modifiedJson, err := json.Marshal(currentSchema)
		if err != nil {
			return VerifyResult{
				Valid: false,
				Issues: []VerifyIssue{{
					Code:    VerifyInternalError,
					Message: fmt.Sprintf("failed to serialize schema at iteration %d", iteration),
				}},
				Error: fmt.Sprintf("marshal (iteration %d): %v", iteration, err),
			}
		}

		resultJson, err := Run(string(modifiedJson), effectiveDate)
		if err != nil {
			return VerifyResult{
				Valid: false,
				Issues: []VerifyIssue{{
					Code:    VerifyInternalError,
					Message: fmt.Sprintf("VM run failed at iteration %d", iteration),
				}},
				Error: fmt.Sprintf("run failed (iteration %d): %v", iteration, err),
			}
		}

		// Parse result
		var resultSchema Schema
		if err := json.Unmarshal([]byte(resultJson), &resultSchema); err != nil {
			return VerifyResult{
				Valid: false,
				Issues: []VerifyIssue{{
					Code:    VerifyInternalError,
					Message: fmt.Sprintf("failed to parse VM result at iteration %d", iteration),
				}},
				Error: fmt.Sprintf("unmarshal result (iteration %d): %v", iteration, err),
			}
		}

		// Build sorted set of visible field IDs for convergence check
		currentVisibleSet := visibleFieldSet(&resultSchema)

		// Check for convergence
		if currentVisibleSet == previousVisibleSet {
			// Converged - now validate the final state and return full result
			return validateFinalState(&newSchema, &resultSchema)
		}

		previousVisibleSet = currentVisibleSet
		currentJson = resultJson
	}

	return VerifyResult{
		Valid: false,
		Issues: []VerifyIssue{{
			Code:    VerifyConvergenceFailed,
			Message: fmt.Sprintf("document did not converge after %d iterations", maxIterations),
		}},
	}
}

// getVisibleEditableFields returns field IDs that are visible and not readonly
func getVisibleEditableFields(schema *Schema) map[string]bool {
	result := make(map[string]bool)
	for id, def := range schema.Definitions {
		if def != nil && def.Visible != nil && *def.Visible && !def.Readonly {
			result[id] = true
		}
	}
	return result
}

// visibleFieldSet returns a sorted string of visible field IDs for convergence checking.
func visibleFieldSet(schema *Schema) string {
	var ids []string
	for id, def := range schema.Definitions {
		if def != nil && def.Visible != nil && *def.Visible {
			ids = append(ids, id)
		}
	}
	sort.Strings(ids)
	return strings.Join(ids, ",")
}

// validateFinalState compares computed values and attestation fulfillment.
// Collects ALL issues instead of bailing on the first — the UI needs the complete picture.
func validateFinalState(newSchema, resultSchema *Schema) VerifyResult {
	engine := &Engine{}
	var issues []VerifyIssue

	// Check for unknown/injected fields in newSchema that don't exist in result
	for id := range newSchema.Definitions {
		if _, existsInResult := resultSchema.Definitions[id]; !existsInResult {
			issues = append(issues, VerifyIssue{
				Code:    VerifyUnknownField,
				FieldID: id,
				Message: fmt.Sprintf("field '%s' does not exist in the schema", id),
			})
		}
	}

	// Compare computed (readonly) values
	for id, resultDef := range resultSchema.Definitions {
		if resultDef == nil || !resultDef.Readonly {
			continue
		}

		newDef, ok := newSchema.Definitions[id]
		if !ok {
			issues = append(issues, VerifyIssue{
				Code:     VerifyComputedMismatch,
				FieldID:  id,
				Message:  fmt.Sprintf("computed field '%s' is missing from the submitted document", id),
				Expected: resultDef.Value,
			})
			continue
		}

		if !engine.compareEqual(newDef.Value, resultDef.Value) {
			issues = append(issues, VerifyIssue{
				Code:     VerifyComputedMismatch,
				FieldID:  id,
				Message:  fmt.Sprintf("computed field '%s' was modified", id),
				Expected: resultDef.Value,
				Claimed:  newDef.Value,
			})
		}
	}

	// Verify attestations are fulfilled
	for id, resultAtt := range resultSchema.Attestations {
		if resultAtt == nil || !resultAtt.Required {
			continue
		}

		newAtt, ok := newSchema.Attestations[id]
		if !ok {
			continue
		}

		if !newAtt.Signed {
			issues = append(issues, VerifyIssue{
				Code:    VerifyAttestationUnsigned,
				FieldID: id,
				Message: fmt.Sprintf("required attestation '%s' has not been signed", id),
			})
			continue // No point checking evidence if unsigned
		}

		if newAtt.Evidence == nil || newAtt.Evidence.ProviderAuditID == "" {
			issues = append(issues, VerifyIssue{
				Code:    VerifyAttestationNoEvidence,
				FieldID: id,
				Message: fmt.Sprintf("attestation '%s' is signed but missing proof of signing", id),
			})
		}

		if newAtt.Evidence == nil || newAtt.Evidence.Timestamp == "" {
			issues = append(issues, VerifyIssue{
				Code:    VerifyAttestationNoTimestamp,
				FieldID: id,
				Message: fmt.Sprintf("attestation '%s' is signed but missing a timestamp", id),
			})
		}
	}

	// Verify status matches
	if newSchema.Status != resultSchema.Status {
		issues = append(issues, VerifyIssue{
			Code:     VerifyStatusMismatch,
			Message:  "the document status does not match what was computed",
			Expected: resultSchema.Status,
			Claimed:  newSchema.Status,
		})
	}

	return VerifyResult{
		Valid:  len(issues) == 0,
		Status: resultSchema.Status,
		Issues: issues,
		Schema: resultSchema,
	}
}

// evaluateLogicTree processes all active rules in order.
func (e *Engine) evaluateLogicTree() {
	for _, rule := range e.schema.LogicTree {
		if rule == nil || rule.Disabled {
			continue
		}

		// Evaluate the condition
		condition := e.resolve(rule.When)
		if e.isTruthy(condition) {
			e.applyAction(rule.Then, rule.ID, rule.LawRef)
		}
	}
}

// applyAction executes a rule's action: setting values, modifying UI, or emitting errors.
func (e *Engine) applyAction(action *Action, ruleID, lawRef string) {
	if action == nil {
		return
	}

	// Apply value mutations
	if action.Set != nil {
		for key, value := range action.Set {
			// Resolve the value in case it's an expression
			resolvedValue := e.resolve(value)
			e.setDefinitionValue(key, resolvedValue, ruleID)
		}
	}

	// Apply UI modifications
	if action.UIModify != nil {
		for key, mods := range action.UIModify {
			e.applyUIModify(key, mods)
		}
	}

	// Emit error if specified
	if action.ErrorMsg != "" {
		kind := action.ErrorKind
		if kind == "" {
			kind = ErrConstraintViolation
		}
		e.addError("", ruleID, kind, action.ErrorMsg, lawRef)
	}
}

// setDefinitionValue updates or creates a definition value.
// Tracks which rule set each field to detect potential cycles.
func (e *Engine) setDefinitionValue(key string, value any, ruleID string) {
	// Cycle detection: check if this field was already set by a different rule
	if prevRule, alreadySet := e.fieldsSet[key]; alreadySet && prevRule != ruleID {
		e.addError(key, ruleID, ErrCycleDetected, fmt.Sprintf(
			"potential cycle: field '%s' set by rule '%s' and again by rule '%s'",
			key, prevRule, ruleID), "")
	}
	e.fieldsSet[key] = ruleID

	def, ok := e.schema.Definitions[key]
	if !ok {
		// Create new definition if it doesn't exist
		t := true
		e.schema.Definitions[key] = &Definition{
			Type:    inferType(value),
			Value:   value,
			Visible: &t,
		}
		return
	}

	def.Value = value
}

// applyUIModify applies UI metadata changes to a definition.
func (e *Engine) applyUIModify(key string, mods any) {
	def, ok := e.schema.Definitions[key]
	if !ok || def == nil {
		return
	}

	modMap, ok := mods.(map[string]any)
	if !ok {
		return
	}

	// Apply visibility and metadata modifications
	if visible, ok := modMap["visible"].(bool); ok {
		def.Visible = &visible
	}
	if uiClass, ok := modMap["ui_class"].(string); ok {
		def.UIClass = uiClass
	}
	if uiMessage, ok := modMap["ui_message"].(string); ok {
		def.UIMessage = uiMessage
	}
	if required, ok := modMap["required"].(bool); ok {
		def.Required = required
	}

	// Apply numeric constraints (min, max, step)
	if minVal, ok := toFloat(modMap["min"]); ok {
		def.Min = &minVal
	}
	if maxVal, ok := toFloat(modMap["max"]); ok {
		def.Max = &maxVal
	}
	if stepVal, ok := toFloat(modMap["step"]); ok {
		def.Step = &stepVal
	}

	// Apply string constraints (min_length, max_length)
	if minLen, ok := modMap["min_length"].(float64); ok {
		intVal := int(minLen)
		def.MinLength = &intVal
	}
	if maxLen, ok := modMap["max_length"].(float64); ok {
		intVal := int(maxLen)
		def.MaxLength = &intVal
	}
	if pattern, ok := modMap["pattern"].(string); ok {
		def.Pattern = pattern
	}
}

// computeDerived evaluates all derived fields in the state model.
func (e *Engine) computeDerived() {
	if e.schema.StateModel == nil || e.schema.StateModel.Derived == nil {
		return
	}

	for name, derivedDef := range e.schema.StateModel.Derived {
		if derivedDef == nil || derivedDef.Eval == nil {
			continue
		}

		// Evaluate the expression
		value := e.resolve(derivedDef.Eval)

		if existing, ok := e.schema.Definitions[name]; ok && existing != nil {
			existing.Value = value
			existing.Readonly = true
			if existing.Visible == nil {
				t := true
				existing.Visible = &t
			}
		} else {
			t := true
			e.schema.Definitions[name] = &Definition{
				Type:     inferType(value),
				Value:    value,
				Readonly: true,
				Visible:  &t,
			}
		}
	}
}

// marshal converts the schema back to JSON.
func (e *Engine) marshal() (string, error) {
	result, err := json.MarshalIndent(e.schema, "", "  ")
	if err != nil {
		return "", fmt.Errorf("marshal: %w", err)
	}
	return string(result), nil
}

// inferType determines the type string for a value.
func inferType(value any) string {
	switch value.(type) {
	case string:
		return "string"
	case float64, int, int64:
		return "number"
	case bool:
		return "boolean"
	default:
		return "string"
	}
}

File: js/src/core/engine.ts
/**
 * Main Tenet VM engine.
 * Provides run() and verify() functions for schema evaluation.
 */

import type {
    TenetSchema,
    TenetResult,
    TenetVerifyResult,
    VerifyIssue,
    EvalState,
    Rule,
    Action,
    Definition,
} from './types.js';
import { resolve } from './resolver.js';
import { isTruthy, toFloat, compareEqual } from './operators.js';
import { applyTemporalRouting } from './temporal.js';
import { validateDefinitions, checkAttestations, determineStatus, addError } from './validate.js';

/**
 * Infer type string from a value.
 */
function inferType(value: unknown): Definition['type'] {
    if (typeof value === 'string') return 'string';
    if (typeof value === 'number') return 'number';
    if (typeof value === 'boolean') return 'boolean';
    return 'string'; // default
}

/**
 * Deep clone an object (for immutability).
 */
function deepClone<T>(obj: T): T {
    return JSON.parse(JSON.stringify(obj));
}

/**
 * Apply UI modifications to a definition.
 */
function applyUIModify(state: EvalState, key: string, mods: unknown): void {
    const def = state.schema.definitions[key];
    if (!def) {
        return;
    }

    if (typeof mods !== 'object' || mods === null) {
        return;
    }

    const modMap = mods as Record<string, unknown>;

    // Apply visibility and metadata modifications
    if (typeof modMap['visible'] === 'boolean') {
        def.visible = modMap['visible'];
    }
    if (typeof modMap['ui_class'] === 'string') {
        def.ui_class = modMap['ui_class'];
    }
    if (typeof modMap['ui_message'] === 'string') {
        def.ui_message = modMap['ui_message'];
    }
    if (typeof modMap['required'] === 'boolean') {
        def.required = modMap['required'];
    }

    // Apply numeric constraints
    const [minVal, minOk] = toFloat(modMap['min']);
    if (minOk) {
        def.min = minVal;
    }
    const [maxVal, maxOk] = toFloat(modMap['max']);
    if (maxOk) {
        def.max = maxVal;
    }
    const [stepVal, stepOk] = toFloat(modMap['step']);
    if (stepOk) {
        def.step = stepVal;
    }

    // Apply string constraints
    if (typeof modMap['min_length'] === 'number') {
        def.min_length = modMap['min_length'];
    }
    if (typeof modMap['max_length'] === 'number') {
        def.max_length = modMap['max_length'];
    }
    if (typeof modMap['pattern'] === 'string') {
        def.pattern = modMap['pattern'];
    }
}

/**
 * Set a definition value, with cycle detection.
 */
function setDefinitionValue(state: EvalState, key: string, value: unknown, ruleId: string): void {
    // Cycle detection
    const prevRule = state.fieldsSet.get(key);
    if (prevRule && prevRule !== ruleId) {
        addError(state, key, ruleId, 'cycle_detected', `Potential cycle: field '${key}' set by rule '${prevRule}' and again by rule '${ruleId}'`);
    }
    state.fieldsSet.set(key, ruleId);

    const def = state.schema.definitions[key];
    if (!def) {
        // Create new definition if it doesn't exist
        state.schema.definitions[key] = {
            type: inferType(value),
            value,
            visible: true,
        };
        return;
    }

    def.value = value;
}

/**
 * Apply a rule's action: setting values, modifying UI, or emitting errors.
 */
function applyAction(state: EvalState, action: Action | undefined, ruleId: string, lawRef: string): void {
    if (!action) {
        return;
    }

    // Apply value mutations
    if (action.set) {
        for (const [key, value] of Object.entries(action.set)) {
            // Resolve the value in case it's an expression
            const resolvedValue = resolve(value, state);
            setDefinitionValue(state, key, resolvedValue, ruleId);
        }
    }

    // Apply UI modifications
    if (action.ui_modify) {
        for (const [key, mods] of Object.entries(action.ui_modify)) {
            applyUIModify(state, key, mods);
        }
    }

    // Emit error if specified
    if (action.error_msg) {
        addError(state, '', ruleId, 'runtime_warning', action.error_msg, lawRef);
    }
}

/**
 * Evaluate the logic tree (all active rules in order).
 */
function evaluateLogicTree(state: EvalState): void {
    if (!state.schema.logic_tree) {
        return;
    }

    for (const rule of state.schema.logic_tree) {
        if (!rule || rule.disabled) {
            continue;
        }

        // Evaluate the condition
        const condition = resolve(rule.when, state);
        if (isTruthy(condition)) {
            applyAction(state, rule.then, rule.id, rule.law_ref || '');
        }
    }
}

/**
 * Compute derived state values.
 */
function computeDerived(state: EvalState): void {
    if (!state.schema.state_model?.derived) {
        return;
    }

    for (const [name, derivedDef] of Object.entries(state.schema.state_model.derived)) {
        if (!derivedDef?.eval) {
            continue;
        }

        // Evaluate the expression
        const value = resolve(derivedDef.eval, state);

        // Preserve existing definition metadata if present
        const existing = state.schema.definitions[name];
        if (existing) {
            existing.value = value;
            existing.readonly = true;
            if (existing.visible === undefined) {
                existing.visible = true;
            }
        } else {
            state.schema.definitions[name] = {
                type: inferType(value),
                value,
                readonly: true,
                visible: true,
            };
        }
    }
}

/**
 * Run the Tenet VM on a schema.
 *
 * @param schema - The schema object
 * @param effectiveDate - Effective date for temporal routing (defaults to now)
 * @returns The transformed schema with computed state, errors, and status
 */
export function run(
    schema: TenetSchema | string,
    effectiveDate: Date | string = new Date()
): TenetResult {
    try {
        // Parse if string
        const parsedSchema: TenetSchema = typeof schema === 'string'
            ? JSON.parse(schema)
            : deepClone(schema);

        // Parse date
        const date = effectiveDate instanceof Date
            ? effectiveDate
            : new Date(effectiveDate);

        // Initialize default visibility for definitions
        for (const def of Object.values(parsedSchema.definitions)) {
            if (def && def.visible === undefined) {
                def.visible = true;
            }
        }

        // Create evaluation state
        const state: EvalState = {
            schema: parsedSchema,
            effectiveDate: date,
            fieldsSet: new Map(),
            errors: [],
            derivedInProgress: new Set(),
        };

        // 1. Select temporal branch and prune inactive rules
        applyTemporalRouting(state);

        // 2. Compute derived state (so logic tree can use derived values)
        computeDerived(state);

        // 3. Evaluate logic tree
        evaluateLogicTree(state);

        // 4. Re-compute derived state (in case logic modified inputs)
        computeDerived(state);

        // 5. Validate definitions
        validateDefinitions(state);

        // 6. Check attestations
        checkAttestations(state, (action, ruleId, lawRef) => {
            applyAction(state, action, ruleId, lawRef);
        });

        // 7. Determine status and attach errors
        state.schema.errors = state.errors.length > 0 ? state.errors : undefined;
        state.schema.status = determineStatus(state);

        return { result: state.schema };
    } catch (error) {
        return { error: String(error) };
    }
}

/**
 * Get visible, editable fields from a schema.
 */
function getVisibleEditableFields(schema: TenetSchema): Set<string> {
    const result = new Set<string>();
    for (const [id, def] of Object.entries(schema.definitions)) {
        if (def && def.visible && !def.readonly) {
            result.add(id);
        }
    }
    return result;
}

/**
 * Get a sorted, comma-joined string of visible field IDs for convergence detection.
 */
function getVisibleFieldIds(schema: TenetSchema): string {
    return Object.entries(schema.definitions)
        .filter(([, def]) => def?.visible)
        .map(([id]) => id)
        .sort()
        .join(',');
}

/**
 * Validate that the final state matches expected values.
 * Collects ALL issues instead of bailing on the first — the UI needs the complete picture.
 */
function validateFinalState(
    newSchema: TenetSchema,
    resultSchema: TenetSchema
): TenetVerifyResult {
    const issues: VerifyIssue[] = [];

    // Check for unknown/injected fields in newSchema that don't exist in result
    for (const id of Object.keys(newSchema.definitions)) {
        if (!(id in resultSchema.definitions)) {
            issues.push({
                code: 'unknown_field',
                field_id: id,
                message: `Field '${id}' does not exist in the schema`,
            });
        }
    }

    // Compare computed (readonly) values
    for (const [id, resultDef] of Object.entries(resultSchema.definitions)) {
        if (!resultDef?.readonly) {
            continue;
        }

        const newDef = newSchema.definitions[id];
        if (!newDef) {
            issues.push({
                code: 'computed_mismatch',
                field_id: id,
                message: `Computed field '${id}' is missing from the submitted document`,
                expected: resultDef.value,
            });
            continue;
        }

        if (!compareEqual(newDef.value, resultDef.value)) {
            issues.push({
                code: 'computed_mismatch',
                field_id: id,
                message: `Computed field '${id}' was modified`,
                expected: resultDef.value,
                claimed: newDef.value,
            });
        }
    }

    // Verify attestations are fulfilled
    if (resultSchema.attestations) {
        for (const [id, resultAtt] of Object.entries(resultSchema.attestations)) {
            if (!resultAtt?.required) {
                continue;
            }

            const newAtt = newSchema.attestations?.[id];
            if (!newAtt) {
                continue;
            }

            if (!newAtt.signed) {
                issues.push({
                    code: 'attestation_unsigned',
                    field_id: id,
                    message: `Required attestation '${id}' has not been signed`,
                });
                continue; // No point checking evidence if unsigned
            }

            if (!newAtt.evidence?.provider_audit_id) {
                issues.push({
                    code: 'attestation_no_evidence',
                    field_id: id,
                    message: `Attestation '${id}' is signed but missing proof of signing`,
                });
            }

            if (!newAtt.evidence?.timestamp) {
                issues.push({
                    code: 'attestation_no_timestamp',
                    field_id: id,
                    message: `Attestation '${id}' is signed but missing a timestamp`,
                });
            }
        }
    }

    // Verify status matches
    if (newSchema.status !== resultSchema.status) {
        issues.push({
            code: 'status_mismatch',
            message: 'The document status does not match what was computed',
            expected: resultSchema.status,
            claimed: newSchema.status,
        });
    }

    return {
        valid: issues.length === 0,
        status: resultSchema.status,
        issues: issues.length > 0 ? issues : undefined,
        schema: resultSchema,
    };
}

/**
 * Verify that a completed document was correctly derived from a base schema.
 * Simulates the user's journey by iteratively copying visible field values and re-running.
 *
 * Returns a structured result with all issues found (not just the first).
 * Crash-safe: catches any unexpected error and returns it as an internal_error issue.
 *
 * @param newSchema - The completed/submitted schema
 * @param oldSchema - The original base schema
 * @param maxIterations - Maximum replay iterations (default: 100)
 * @returns Structured verification result
 */
export function verify(
    newSchema: TenetSchema | string,
    oldSchema: TenetSchema | string,
    maxIterations: number = 100
): TenetVerifyResult {
    try {
        // Parse schemas
        const parsedNewSchema: TenetSchema = typeof newSchema === 'string'
            ? JSON.parse(newSchema)
            : deepClone(newSchema);

        const parsedOldSchema: TenetSchema = typeof oldSchema === 'string'
            ? JSON.parse(oldSchema)
            : deepClone(oldSchema);

        // Extract effective date from newSchema
        let effectiveDate = new Date();
        if (parsedNewSchema.valid_from) {
            const parsed = new Date(parsedNewSchema.valid_from);
            if (!isNaN(parsed.getTime())) {
                effectiveDate = parsed;
            }
        }

        // Start with base schema
        let currentSchema = parsedOldSchema;
        let previousVisibleIds = '';

        for (let iteration = 0; iteration < maxIterations; iteration++) {
            // Count visible editable fields before copying
            const visibleEditable = getVisibleEditableFields(currentSchema);

            // Copy values from newSchema for visible, editable fields
            for (const fieldId of visibleEditable) {
                const newDef = parsedNewSchema.definitions[fieldId];
                const currentDef = currentSchema.definitions[fieldId];
                if (newDef && currentDef) {
                    currentDef.value = newDef.value;
                }
            }

            // Copy attestation states
            if (currentSchema.attestations) {
                for (const [attId, currentAtt] of Object.entries(currentSchema.attestations)) {
                    if (!currentAtt) continue;
                    const newAtt = parsedNewSchema.attestations?.[attId];
                    if (newAtt) {
                        currentAtt.signed = newAtt.signed;
                        currentAtt.evidence = newAtt.evidence;
                    }
                }
            }

            // Run the schema
            const runResult = run(currentSchema, effectiveDate);
            if (runResult.error) {
                return {
                    valid: false,
                    issues: [{
                        code: 'internal_error',
                        message: `VM run failed at iteration ${iteration}`,
                    }],
                    error: `Run failed (iteration ${iteration}): ${runResult.error}`,
                };
            }

            const resultSchema = runResult.result!;

            // Get visible field IDs after run
            const currentVisibleIds = getVisibleFieldIds(resultSchema);

            // Check for convergence using set comparison
            if (currentVisibleIds === previousVisibleIds) {
                // Converged - now validate the final state and return full result
                return validateFinalState(parsedNewSchema, resultSchema);
            }

            previousVisibleIds = currentVisibleIds;
            currentSchema = resultSchema;
        }

        return {
            valid: false,
            issues: [{
                code: 'convergence_failed',
                message: `Document did not converge after ${maxIterations} iterations`,
            }],
        };
    } catch (error) {
        return {
            valid: false,
            issues: [{
                code: 'internal_error',
                message: `Unexpected error during verification`,
            }],
            error: String(error),
        };
    }
}
